### PR TITLE
Allows token counts to be streamed out of the app

### PR DESCRIPTION
**Pull request by ghost**
## Context

While token counting has been implemented as part of the metadata, the data is not being correctly streamed out. This fixes the problem and adds unit tests to check it works.

## Changes proposed in this pull request

* Updates `to_request_metadata()` to dispatch custom events when metadata is generated
* Adds a `build_llm_chain()` that formalises some of the slightly weird LangChain needed to use this function
* Adds or updates unit tests for both functions, and the app itself
* Minor updates to the `metadata_reducer()` so it'll handle big sends, and works in the unit tests  

## Guidance to review

* Check you're happy with the addition of `build_llm_chain()`. It give me a bit of code smell and I'm not sure why, but given how fiddly metadata generation is, it makes sense, I think
* Check you're happy with how I handle token counting for different models

## Relevant links

* [How to dispatch custom callback events](https://python.langchain.com/v0.2/docs/how_to/callbacks_custom_events/#sync-callback-handler)

## Things to check

- [x] I have added any new ENV vars in all deployed environments
- [x] I have tested any code added or changed
- [x] [I have run integration tests](https://github.com/i-dot-ai/redbox/actions/runs/10593533651)

### CODE REVIEW

File: redbox-core/redbox/chains/runnables.py [hunk 1]
```diff
@@ -65,6 +67,18 @@ def _chat_prompt_from_messages(state: RedboxState) -> Runnable:
     return _chat_prompt_from_messages
 
 
+def build_llm_chain(prompt_set: PromptSet, llm: BaseChatModel) -> Runnable:
+    """Builds a chain that correctly a text and metadata state update.
```
Comment (rachaelcodes) [inline redbox-core/redbox/chains/runnables.py:71]: think there's a typo here?

File: redbox-core/redbox/graph/nodes/processes.py [hunk 2]
```diff
@@ -136,14 +121,9 @@ def build_stuff_pattern(
...
         _llm = llm.with_config(tags=["response_flag"]) if final_response_chain else llm
-        return (
-            build_chat_prompt_from_messages_runnable(prompt_set)
-            | _llm
-            | {
-                "text": StrOutputParser(),
-                "metadata": RunnableLambda(attrgetter("response_metadata")) | to_request_metadata,
-            }
-        ).invoke(state)
+        stuff_chain = build_llm_chain(prompt_set=prompt_set, llm=_llm)
```
Comment (rachaelcodes) [inline redbox-core/redbox/graph/nodes/processes.py:124]: Do we need to declare this variable?

File: redbox-core/redbox/chains/runnables.py [hunk 3]
```diff
@@ -65,6 +67,18 @@ def _chat_prompt_from_messages(state: RedboxState) -> Runnable:
     return _chat_prompt_from_messages
 
 
+def build_llm_chain(prompt_set: PromptSet, llm: BaseChatModel) -> Runnable:
```
Comment (gecBurton) [inline redbox-core/redbox/chains/runnables.py:70]: nice bit of DRY

File: redbox-core/redbox/transform.py [hunk 4]
```diff
@@ -61,21 +65,18 @@ def flatten_document_state(documents: DocumentState) -> list[Document]:
     return [document for group in documents.values() for document in group.values()]
 
 
-def to_request_metadata(response_metadata: dict):
-    """
-    Convert a Langchain ChatLLM response metadata dictionary
+@RunnableLambda
+def to_request_metadata(output: dict):
```
Comment (gecBurton) [inline redbox-core/redbox/transform.py:69]:
```suggestion
def to_request_metadata(output: dict) -> RequestMetadata:
```

nitpick: `output` feels like a strange variable name for an input?

File: core-api/core_api/routes/chat.py [hunk 5]
```diff
@@ -103,7 +103,7 @@ async def on_documents_available(docs: list[Document]):
             websocket,
         )
 
-    async def on_metadata_available(metadata: RequestMetadata):
+    async def on_metadata_response(metadata: RequestMetadata):
```
Comment (rachaelcodes) [inline core-api/core_api/routes/chat.py:106]: nice renaming here

File: redbox-core/redbox/transform.py (hunk 4 above)
Comment (gecBurton) [reply redbox-core/redbox/transform.py:69]: otherwise nice and clear

File: redbox-core/tests/graph/test_app.py [hunk 6]
```diff
@@ -158,18 +159,23 @@ async def test_chat(test: RedboxChatTestCase, env, mocker):
...
 
     final_state = RedboxState(response)
+
     assert (
         final_state["text"] == test_case.test_data.expected_llm_response[-1]
     ), f"Expected LLM response: '{test_case.test_data.expected_llm_response[-1]}'. Received '{final_state["text"]}'"
     assert (
         final_state["route_name"] == test_case.test_data.expected_route
     ), f"Expected Route: '{ test_case.test_data.expected_route}'. Received '{final_state["route_name"]}'"
+    if final_state.get("metadata"):
```
Comment (gecBurton) [inline redbox-core/tests/graph/test_app.py:176]: nitpick

```suggestion
    if metadata:=final_state.get("metadata"):
```

File: redbox-core/tests/graph/test_app.py [hunk 7]
```diff
@@ -158,18 +159,23 @@ async def test_chat(test: RedboxChatTestCase, env, mocker):
...
+        assert final_state["metadata"].get("input_tokens", 0) > 0
```
Comment (gecBurton) [inline redbox-core/tests/graph/test_app.py:177]: nitpick:

```suggestion
        assert final_state["metadata"].get("input_tokens", 0)
```

**Review by gecBurton — APPROVED** [review]

**Review by jamesrichards4 — APPROVED** [review]

**Comment by jamesrichards4** [comment]
@rachaelcodes It looks to me like our conversation on structure of these token counts is resolved by this work? We have counts by model in the final metadata object but the Django app could also build up a list of all events from the dispatched metadata events like we were discussing too. Not sure which is best tbh but options can't be a bad thing

**Pull request merged**